multidownloader/storage/storage.go

```diff
@@ -0,0 +1,396 @@
+package storage
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	aggkitcommon "github.com/agglayer/aggkit/common"
+	"github.com/agglayer/aggkit/db"
+	dbtypes "github.com/agglayer/aggkit/db/types"
+	"github.com/agglayer/aggkit/multidownloader/storage/migrations"
+	mdrtypes "github.com/agglayer/aggkit/multidownloader/types"
+	aggkittypes "github.com/agglayer/aggkit/types"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/jmoiron/sqlx"
+	"github.com/russross/meddler"
+)
+
+type MultidownloaderStorageConfig struct {
+	DBPath string
+}
+
+type MultidownloaderStorage struct {
+	dbtypes.KeyValueStorager
+	logger aggkitcommon.Logger
+	db     *sql.DB
+	cfg    MultidownloaderStorageConfig
+
+	mutex sync.RWMutex
+}
+
+type logRow struct {
+	Address     common.Address `meddler:"address,address"`
+	Topics      string         `meddler:"topics"`
+	Data        []byte         `meddler:"data"`
+	BlockNumber uint64         `meddler:"block_number"`
+	TxHash      common.Hash    `meddler:"tx_hash,hash"`
+	TxIndex     uint           `meddler:"tx_index"`
+	Index       uint           `meddler:"log_index"`
+}
+
+func (l *logRow) String() string {
+	if l == nil {
+		return "logRow{nil}"
+	}
+	return fmt.Sprintf("logRow{Address: %s, Topics: %s, DataLen: %d, BlockNumber: %d, "+
+		"TxHash: %s, TxIndex: %d, Index: %d}",
+		l.Address.Hex(), l.Topics, len(l.Data), l.BlockNumber, l.TxHash.Hex(), l.TxIndex, l.Index)
+}
+
+func NewLogRowsFromEthLogs(logs []types.Log) []*logRow {
+	rows := make([]*logRow, 0, len(logs))
+	for _, log := range logs {
+		row := NewLogRowFromEthLog(log)
+		rows = append(rows, row)
+	}
+	return rows
+}
+
+type syncStatusRow struct {
+	Address         common.Address `meddler:"contract_address,address"`
+	TargetFromBlock uint64         `meddler:"target_from_block"`
+	TargetToBlock   string         `meddler:"target_to_block"`
+	SyncedFromBlock uint64         `meddler:"synced_from_block"`
+	SyncedToBlock   uint64         `meddler:"synced_to_block"`
+	SyncersIDs      string         `meddler:"syncers_id"`
+}
+
+func NewLogRowFromEthLog(log types.Log) *logRow {
+	topicsJSON, err := json.Marshal(log.Topics)
+	if err != nil {
+		// If marshaling fails, fallback to empty string or handle error as needed
+		topicsJSON = []byte("[]")
+	}
+	return &logRow{
+		Address:     log.Address,
+		Topics:      string(topicsJSON),
+		Data:        log.Data,
+		BlockNumber: log.BlockNumber,
+		TxHash:      log.TxHash,
+		TxIndex:     log.TxIndex,
+		Index:       log.Index,
+	}
+}
+
+const SqliteBoolTrue = 1
+const SqliteBoolFalse = 0
+
+type BlockRow struct {
```

Comment:
This doesn't look like it have to be exported.
```suggestion
type blockRow struct {
```

Reply:
d49e981de0e722508d4972ae41daf2e029231f26